src/page-storage/store-page.js

```diff
@@ -1,66 +1,97 @@
-    return {finalPage}
+    return finalPage
 }
 
-// Tries to reidentify the page in the tab, or creates a new page doc for it.
+// Tries to reidentify the page in tab, or creates a new page doc for it.
 // Either way, it returns the page doc. If a new page was created, it starts
-// page analysis and then deduplication, and also returns a promise
-// (finalPagePromise) of the page doc resulting from these steps.
-export async function reidentifyOrStorePage({tabId, url}) {
+// page analysis and then deduplication, and also returns an async function
+// (fetchFinalPage) returning the page doc resulting from these steps.
+export async function reidentifyOrStorePage({ tabId, url }) {
     // Find pages we know that had the same URL.
-    const samePageCandidates = (await findPagesByUrl({url})).rows.map(row => row.doc)
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
 
     // Check if we can already tell in advance that this is the same page.
-    const reusablePage = await tryReidentifyPage({tabId, url, samePageCandidates})
+    const reusablePage = await tryReidentifyPage({ tabId, url, samePageCandidates })
 
     if (reusablePage) {
         // Equality is known (or assumed) in advance. Reuse the old page as is.
-        return {page: reusablePage}
+        return { page: reusablePage }
     } else {
-        // Create a new page doc in the database.
-        const page = await createPageStub({url})
-
-        // Start analysis and (possibly) deduplication, but do not wait for it.
-        const finalPagePromise = analyseAndTryDedupePage({
-            tabId,
-            page,
-            samePageCandidates,
-        })
-
-        // Return the page stub, and a promise of the analysed & deduped page.
-        return {page, finalPagePromise}
+        // Create a new page doc stub in the database.
+        const page = await storePage(createPage({ url }))
+
+        // Set up fetching function to run anaylsis + fetching logic on-demand
+        const fetchFinalPage = () => analyseAndTryDedupePage({ tabId, page, samePageCandidates })
+
+        // Return the page stub, and a function to get the analysed & deduped page.
+        return { page, fetchFinalPage }
+    }
+}
+
+/**
+ * Given a URL, attempts to perform all logic needed to store a full page document.
+ * @param {string} url The URL to attempt to store a page document for.
+ * @returns {any} Page document that was created (or found, if reusable/deduped).
+ */
+export async function storePageFromUrl(url) {
+    // Find pages we know that had the same URL.
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
+
+    // Check if we can already tell in advance that this is the same page.
+    const reusablePage = await tryReidentifyPage({ url, samePageCandidates })
+
+    // Equality is known (or assumed) in advance. Reuse the old page as is.
+    if (reusablePage) {
+        return reusablePage
     }
```

Comment:
To give you more of an idea of the batcher in use with `storePageFromUrl `, I have played with rewriting the existing `import-history` to be based on the batcher in [this commit](https://github.com/WorldBrain/WebMemex/commit/520074884ac87334137fdcc2b2c7a7500d7aaf8e) on WorldBrain branch where I'm working on hooking things up.

It's not final but I'm planning to init the batcher here in the background and make some interface to it, via the RPC wrapper lib that we have, to call and manage from the UI redux actions. Main things is the observer functions (`next run after each input processed, `error` run after any errors cancel out the input processing, and `complete` run after all input is processed or cancelled). You can see reindexing, and visit page generation and storage is run after the page batch complete (visits are derived from pages, and needs all pages to do the inter-visit doc referencing).